retrolambda support & cleaner type detection for lambda's

csoroiu opened the pull request:
Added support for `retrolambda` library:
- The lambda's created by this library add a field named `instance` inside the generated classes). I will try to figure out how I can create a test that imitates this behavior. 
- I removed the version restrictions for the lambda support.

Also I did a refactor and made the code work without the `TypeDescriptor` class. Not sure how to test android compatibility, but I expect it to work.

csoroiu commented:
re-opening. Apparently there was an issue when I used retro lambda and jacoco in the same project, but is issue has nothing to do with this change.

jhalterman commented on src/main/java/net/jodah/typetools/TypeResolver.java:
```diff
@@ -250,7 +239,7 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
 
     if (map == null) {
-      map = new HashMap<TypeVariable<?>, Type>();
+      map = new HashMap<>();
```
TypeTools compilation is currently targeted at 1.7, so I don't think we can use the diamond operator.

csoroiu replied:
I realized that later and the next commit #b7a7fa6  fixes the diamond issue.

jhalterman commented on src/main/java/net/jodah/typetools/TypeResolver.java:
```diff
@@ -282,7 +271,7 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
 
       if (CACHE_ENABLED)
-        typeVariableCache.put(targetType, new WeakReference<Map<TypeVariable<?>, Type>>(map));
+        typeVariableCache.put(targetType, new WeakReference<>(map));
```
Same here regarding the diamond operator.

jhalterman commented on src/main/java/net/jodah/typetools/TypeResolver.java:
```diff
@@ -296,7 +285,7 @@ private static void populateLambdaArgs(Class<?> functionalInterface, final Class
       // Find SAM
       for (Method m : functionalInterface.getMethods()) {
-        if (!m.isDefault() && !Modifier.isStatic(m.getModifiers()) && !m.isBridge()) {
+        if (!isDefaultMethod(m) && !Modifier.isStatic(m.getModifiers()) && !m.isBridge()) {
```
Not sure why this change was necessary?

csoroiu commented:
I added that check because I removed the version check at the class initialization and It would fail with NoSuchMethodError when ran with java 6 or 7.

jhalterman commented:
Looks good - thanks for this!

jhalterman commented:
0.4.8 is released, with this change.

csoroiu commented:
Thanks a lot.
